lib/topics.js

```diff
@@ -8,18 +8,37 @@ const SNS    = require('./sns');
 class Topic {
 
   constructor(server, topicName) {
-    this._server = server;
-    this.name    = topicName;
+    this._server     = server;
+    this.name        = topicName;
+    this.subscribers = [];
   }
 
 
   // Publish to a topic.
   publish(message) {
-    return this.clientPromise
-      .then(client => {
-        const stringMessage = JSON.stringify(message);
-        return client.publish(this.name, stringMessage);
+    const stringMessage = JSON.stringify(message);
+
+    if (process.env.NODE_ENV === 'test') {
+      // JSON encode/decode the message before passing to
+      // subscriber, same transformation as when using SNS.
+      const parsedMessage = JSON.parse(stringMessage);
+      setImmediate(() => {
+        this.subscribers.forEach(handler => handler(parsedMessage));
       });
+      return Promise.resolve();
+    } else {
+      return this.clientPromise
+        .then(client => client.publish(this.name, stringMessage));
+    }
```

Comment:
The way I imagine this working:

* Application can subscribe/unsubscribe at will, as often as it needs to
* Internally, there’s a method that dispatches the notification to all subscribers, and does so asynchronously, concurrently, and with error handling (let Ironium log and report)
* During testing, publish calls dispatch, so all notifications handled internally
* In production, publish would use the message broker API (e.g SNS) to publish
* There would be a separate mechanism for receiving these notifications (e.g webhook) and calling dispatch
* The receive function would verify the authenticity of the notification, unpack it, call the dispatch function
* When the topic is configured statically (always exists), then the receiving endpoint needs to be known in advance, and use a request handler to receive and dispatch each notification
* When the topic is configured dynamically, the the topic would use the broker API to add a subscription when the first subscriber is added
* Which one we use depends on the provider, i.e there’s a performance concern if adding a subscriber is a lengthy operation (each process would attempt that at run time)